### DISCUSSION TITLE
How to better understand memory?

### DISCUSSION THREAD

**Discussion by EWCunha** [Cyfrin/assembly-evm-opcodes-and-formal-verification-course#12] [Q&A] [answered]
If I'm not mistaken, Patrick explains the memory as an array-like structure that stores non-persistent data, right? I understood the only difference between storage and memory is that the former is persistent while the latter isn't. So I thought that a parameter of MSTORE would be a memory slot to store the value just like SSTORE instead of a memory offset.

Then, maybe it is better to understand memory as sort of a string-like structure, in which I need to provide the index (offset) where is the starting point to store some data in memory?

I just want to be sure that I got it right...

Thank you

**Comment by EngrPips** [comment]
Hello @EWCunha, Understanding EVM Memory was quite hard for me because of the offset of a thing. Still, it got more manageable when I started thinking about storage as an array of 32 bytes demarcated and Mmeory as an array of 32 bytes not demarcated. So because the array of 32 bytes in memory is not demarcated, nothing stops you from overflowing from one array of 32 bytes to another array of 32 bytes. Hence, I understand why we must constantly provide an offset to declare clearly where we want to start writing. Honestly, I could have explained it better, so @PatrickAlphaC  will help you understand better whenever they are available.

**Comment by PatrickAlphaC** [comment, marked as answer]
@EWCunha you're right, basically!

Here are the inputs for `MSTORE`
```
- offset: offset in the memory in bytes.
- value: 32-byte value to write in the memory.
```

And here are the inputs for `SSTORE`

```
- key: 32-byte key in storage.
- value: 32-byte value to store.
```

The `key` and `offset` are essentially the same, with a slight difference. 

Storage is more strict with "slots" whereas memory is more about bytes offset. 

## SSTORE example

```
PUSH1 0x01
PUSH1 0x01
SSTORE
```

This will result in storage looking like this:

```
0x00: 0x00...00
0x01: 0x00...01
```

## MSTORE example

However, memory will look a little different


```
PUSH1 0x01
PUSH1 0x01
MSTORE
```

```
0x00: 0x00
0x01: 0x00        <---starts placing the 0x01, padded to be 32 bytes long
0x02: 0x00
0x03: 0x00
.
.
.
0x1e: 0x00
0x1f: 0x00
0x20: 0x01
0x21: 0x00
```


## Explained

For storage, every storage slot is a 32 byte slot. So slot 0 has 32 bytes. Slot 1 has 32 bytes.

For memory, every memory slot is a 1 byte slot.

Hopefully that clears it up!

**Reply by PatrickAlphaC** [reply]
Try these in the https://www.evm.codes/playground playground to see the outcomes :)

**Reply by EngrPips** [reply]
Thanks for all you do sir.

**Reply by EWCunha** [reply]
Thank you very much @PatrickAlphaC this really cleared everything!
